overlay/generic/smartdc/lib/smartos_prompt_config.sh

```diff
@@ -141,52 +222,221 @@ is_net()
 	[ $c -lt 0 ] && return 1
 	[ $c -gt 255 ] && return 1
 	[ $d -lt 0 ] && return 1
-	# Make sure the last field isn't the broadcast addr.
-	[ $d -ge 255 ] && return 1
+	return 0
+}
+
+# Tests if input is an email address
+is_email() {
+	regex="^[a-z0-9!#\$%&'*+/=?^_\`{|}~-]+(\.[a-z0-9!#$%&'*+/=?^_\`{|}~-]+)*@([a-z0-9]([a-z0-9-]*[a-z0-9])?\.?)+[a-z0-9]([a-z0-9-]*[a-z0-9])?\$"
+	ADDRESS=$1
+
+	[[ $ADDRESS =~ $regex ]] && return 0
+	return 1
+}
+
+# You can call this like:
+#
+#  value=$(getanswer "foo")
+#  [[ $? == 0 ]] || fatal "no answer for question foo"
+#
+getanswer()
+{
+	local key=$1
+	local answer=""
+	local potential=""
+
+	if [[ -z ${answer_file} ]]; then
+		return ${ENOTFOUND}
+	fi
+
+	# json does not distingush between an empty string and a key that's not
+	# there with the normal output, so we fix that so we can distinguish.
+	answer=$(/usr/bin/cat ${answer_file} \
+		| /usr/bin/json -e "if (this['${key}'] === undefined) this['${key}'] = '<<undefined>>';" \
+		"${key}" 2>&1)
+	if [[ $? != 0 ]]; then
+		if [[ -n $(echo "${answer}" | grep "input is not JSON") ]]; then
+			return ${EBADJSON}
+		else
+			return ${EUNKNOWN}
+		fi
+	fi
+
+	if [[ ${answer} == "<<undefined>>" ]]; then
+		return ${ENOTFOUND}
+	fi
+
+	echo "${answer}"
 	return 0
 }
 
 # Optional input
 promptopt()
 {
-	val=
-	printf "%s [press enter for none]: " "$1"
+	val=""
+	def="$2"
+	key="$3"
+
+	if [[ -n ${key} ]]; then
+		val=$(getanswer "${key}")
+		if [[ $? == 0 ]]; then
+			if [[ ${val} == "<default>" ]]; then
+				val=${def}
+			fi
+			return
+		fi
+	fi
+
+	if [ -z "$def" ]; then
```

Comment:
personally, i'd prefer `[[` over `[` as it is a builtin and more robust than `[`... see http://mywiki.wooledge.org/BashFAQ/031 for more info

also, quotes around `$def` are not needed if `[[ ... ]]` is used.